src/sync.cpp

```diff
@@ -1276,5 +1312,154 @@ bool Sync::movetolocaldebris(string* localpath)
 
     return false;
 }
+
+void Sync::invalidatefsids(LocalNode& l)
+{
+    l.fsid = ~handle{0};
+    if (l.fsid_it != client->fsidnode.end())
+    {
+        client->fsidnode.erase(l.fsid_it);
+        l.fsid_it = client->fsidnode.end();
+    }
+    if (l.type == FILENODE)
+    {
+        return;
+    }
+    for (auto it = l.children.begin(); it != l.children.end(); ++it)
+    {
+        invalidatefsids(*it->second);
+    }
+}
+
+void Sync::assignfsids(bool& success, string& localpath, std::unique_ptr<FileAccess> fa,
+                       const std::multiset<FileFingerprint*, FileFingerprintCmp>& nodes)
+{
+    if (!success)
+    {
+        return;
+    }
+
+    if (!isscannable(localpath))
+    {
+        return;
+    }
+
+    auto da = std::unique_ptr<DirAccess>{client->fsaccess->newdiraccess()};
+    if (!(success = da->dopen(&localpath, fa.get(), false)))
+    {
+        LOG_warn << "Unable to open directory: " << localpath;
+        return;
+    }
+
+    string localname;
+    const size_t localpathSize = localpath.size();
+
+    while (da->dnext(&localpath, &localname, client->followsymlinks))
+    {
+        auto name = localname;
+        client->fsaccess->local2name(&name);
+
+        if (localpathSize > 0)
+        {
+            localpath.append(client->fsaccess->localseparator);
+        }
+
+        localpath.append(localname);
+
+        // check if this record is to be ignored
+        if (client->app->sync_syncable(this, name.c_str(), &localpath))
+        {
+            // skip the sync's debris folder
+            if (isscannable(localpath))
+            {
+                auto localfa = std::unique_ptr<FileAccess>{client->fsaccess->newfileaccess()};
+                if ((success = localfa->fopen(&localpath, true, false)))
+                {
+                    if (localfa->type == FILENODE)
+                    {
+                        success = assignfsid(*localfa, localpath, nodes);
+                    }
+                    else if (localfa->type == FOLDERNODE)
```

Comment:
shouldn't we assign fsids to localnodes representing folders also, so we can detect folder moves?  Though, it might be easier to do so by walking the localnode tree rather than walking the filesystem

Reply:
We cannot assign IDs to folders because we don't know which ID to assign. Folders don't have a full fingerprint so we cannot construct any heuristic to map fingerprints to IDs like we do with files. Does that make sense?